2-virtual-module-source.md

```diff
@@ -0,0 +1,397 @@
+# Virtual Module Sources
+
+## Synopsis
+
+Extend the `Module` constructor such that it accepts virtual module sources:
+objects that implement a protocol that is sufficient for virtualizing the
+evaluation of modules in languages not anticipated by ECMA-262 or a host
+implementations.
+
+## Motivation
+
+Hosts can add support for linking non-EcmaScript languages by creating
+new kinds of ***Module Source Record***.
+For example, a host could add a [[ModuleSource]] internal slot to
+[`WebAssembly.Module`][web-assembly-module] and then link arbitrary WASM into
+an EcmaScript module graph.
+
+However, user code cannot emulate a host in this way.
+To extend or experiment with new languages or new module-like features, user
+code needs a way to emulate or *virtualize* a module source record.
+
+Such a protocol would allow user code to integrate CommonJS, JSON, or WASM
+modules on hosts that do not.
+It would also open a field for experimentation with other kinds of resource
+modules.
+
+## Description
+
+The first argument to the `Module` constructor is a source.
+If that source is an object that does not have a [[Module Source]] internal
+slot, we treat this object as a protocol for loading, binding, linking,
+initializing, and executing the new `Module` instance.
+
+## Interface
+
+```ts
+type VirtualModuleSource = {
```

Comment:
Side Note: I strongly believe that this is not the best place, and instead what you want is really a non-transferable (or non-serializable) structure, which ModuleSource is not. I believe a far better choice would be to have a `VirtualModule` instead, which is going to be far lower level control mechanism, but a lot more powerful and simpler. It might be that you have a combination of `VirtualModuleSource` + `VirtualModule` to match the existing `ModuleSource` + `Module`, where one is meta and it is serializable, and the other one is not.

Reply:
In the case of the `VirtualModule`, all we need is a way to allow setting values of a list of predefined slots (described by `VirtualModuleSource` by exported name bindings), which is a well known mechanism used by modules today where the value must be set into the environment record associated to a module.

Again, we go from spec'ing a "parameterized artifact" to "parameterized low-level APIs" so developers can create the artifact themself.